File: jgl/testExperiment.js
/**
 * 
 */
$(document).ready(function() {
	testExperiment();
});

function testExperiment() {
	window.myscreen = initScreen();
	
	window.task = [];
	task[0] = [];
	task[0][0] = {};
	task[0][0] = initSurvey();
	task[0][1] = {};
	task[0][1].seglen = [2, Infinity];
	task[0][1].getResponse = [1];
	task[0][1].numBlocks = 10;
	task[0][1].numTrials = 10;
	task[0][1].parameter = {};
	task[0][1].parameter.dir = 0;
	task[0][1].parameter.coherence = 0;
	task[0][1].html = "canvas.html";
	task[0][1].usingScreen = 1;
	task[0][1] = initTask(task[0][1], startSegmentCallback, updateScreenCallback);
	task[0][2] = initSurvey(myscreen);
//	task[0][2].html = "postquestionnaire.html";
	
	window.stimulus = {};
	
	initStimulus("stimulus");
	task[0][1] = initDots(task[0][1], myscreen);
	
	initTurk();
	
	startPhase(task[0]);
	
}


var startSegmentCallback = function(task, myscreen) {
	if (! window.stimulus.hasOwnProperty("dots")) {
		window.stimulus.dots = {};
	}
	
	if (task.thistrial.thisseg == 1) {
		window.stimulus.dots.coherence = task.thistrial.coherence;
	} else {
		window.stimulus.dots.coherence = 0;
	}
	
	window.stimulus.dots.dir = task.thistrial.dir;
	
	console.log(jglGetSecs());

	return [task, myscreen];
}

var updateScreenCallback = function(task, myscreen) {
	jglClearScreen();
	window.stimulus = updateDots(window.stimulus, task, myscreen);
	return [task, myscreen];
}

function initDots(task, myscreen) {
	if (! stimulus.hasOwnProperty("dots")) {
		stimulus.dots = {};
	}
	
	if (! stimulus.dots.hasOwnProperty("rmax")) {
		stimulus.dots.rmax = 5;
	}
	
	if (! stimulus.dots.hasOwnProperty("xcenter")) {
		stimulus.dots.xcenter = 0;
	}
	
	if (! stimulus.dots.hasOwnProperty("ycenter")) {
		stimulus.dots.ycenter = 0;
	}
	
	if (! stimulus.dots.hasOwnProperty("dotsize")) {
		stimulus.dots.dotsize = 0.2;
	}
	
	if (! stimulus.dots.hasOwnProperty("density")) {
		stimulus.dots.density = 0.3;
	}
	
	if (! stimulus.dots.hasOwnProperty("coherence")) {
		stimulus.dots.coherence = 1;
	}
	if (! stimulus.dots.hasOwnProperty("speed")) {
		stimulus.dots.speed = 5;
	}
	if (! stimulus.dots.hasOwnProperty("dir")) {
		stimulus.dots.dir = 0;
	}
	
	stimulus.dots.width = 10;
	stimulus.dots.height = 10;
	
	stimulus.dots.n = Math.round(stimulus.dots.width * stimulus.dots.height * stimulus.dots.density);
	
	stimulus.dots.xmin = -stimulus.dots.width / 2;
	stimulus.dots.xmax = stimulus.dots.width / 2;
	stimulus.dots.ymin = -stimulus.dots.height / 2;
	stimulus.dots.ymax = stimulus.dots.height / 2;
	
	stimulus.dots.x = multiply(rand(task, stimulus.dots.n), stimulus.dots.width);
	stimulus.dots.y = multiply(rand(task, stimulus.dots.n), stimulus.dots.height);
	
	stimulus.dots.x = subtract(stimulus.dots.x, stimulus.dots.width / 2);
	stimulus.dots.y = subtract(stimulus.dots.y, stimulus.dots.height / 2);
	
	stimulus.dots.stepsize = stimulus.dots.speed / myscreen.framesPerSecond;
	
	jglOpen(myscreen.ppi);
	
	jglClearScreen();
	
	jglStencilCreateBegin(1);
	
	jglPoints2([stimulus.dots.xcenter], [stimulus.dots.ycenter], stimulus.dots.rmax, "#ffffff");
	
	jglStencilCreateEnd();
	
	jglClearScreen();
	
	jglClose();
	
	return task;
}

function updateDots(stimulus, task, myscreen) {
//	stimulus.dots.xstep = Math.cos(Math.PI * stimulus.dots.dir / 180) * stimulus.dots.stepsize;
//	stimulus.dots.ystep = Math.sin(Math.PI * stimulus.dots.dir / 180) * stimulus.dots.stepsize;
//	
//	stimulus.dots.coherent = lessThan(rand(task, stimulus.dots.n), stimulus.dots.coherence);
//	
//	stimulus.dots.x = add(stimulus.dots.x, stimulus.dots.xstep, stimulus.dots.coherent);
//	stimulus.dots.y = add(stimulus.dots.y, stimulus.dots.ystep, stimulus.dots.coherent);
//	
//	var notCo = not(stimulus.dots.coherent);
//	var thisdir = rand(task, sum(notCo)) * 2 * Math.PI;
//	
//	stimulus.dots.x = add(stimulus.dots.x, multiply(cos(thisdir), stimulus.dots.stepsize), notCo);
//	stimulus.dots.y = add(stimulus.dots.y, multiply(sin(thisdir), stimulus.dots.stepsize), notCo);
//	
//	stimulus.dots.x = add(stimulus.dots.x, stimulus.dots.width, lessThan(stimulus.dots.x, stimulus.dots.xmin));
//	stimulus.dots.x = subtract(stimulus.dots.x, stimulus.dots.width, greaterThan(stimulus.dots.x, stimulus.dots.xmax));
//	stimulus.dots.y = add(stimulus.dots.y, stimulus.dots.height, lessThan(stimulus.dots.y, stimulus.dots.ymin));
//	stimulus.dots.y = subtract(stimulus.dots.y, stimulus.dots.height, greaterThan(stimulus.dots.y, stimulus.dots.ymax));
	
	stimulus.dots.x = add(stimulus.dots.x,0.1);

	var over = greaterThan(stimulus.dots.x, 10);
	stimulus.dots.x = subtract(stimulus.dots.x, 10, over);
	
	jglStencilSelect(1);
	jglPoints2(stimulus.dots.x, stimulus.dots.y, stimulus.dots.dotsize, "#000000");
	jglStencilSelect(0);

	return stimulus;

}

File: task/testExperiment.m
% testExperiment.m
%
%        $Id$
%      usage: testExperiment
%         by: justin gardner
%       date: 09/07/06
%  copyright: (c) 2006 Justin Gardner (GPL see mgl/COPYING)
%    purpose: example program to show how to use the task structure
%
function myscreen = testExperiment(screenName)

% check arguments
if ~any(nargin == [0 1])
  help testExperiment
  return
end

% initalize the screen
if nargin >= 1
  myscreen = initScreen(screenName);
else
  myscreen = initScreen;
end

% set the first task to be the fixation staircase task
[task{1} myscreen] = fixStairInitTask(myscreen);

% set our task to have two phases. 
% one starts out with dots moving for incohrently for 10 seconds
task{2}{1}.waitForBacktick = 1;
task{2}{1}.seglen = 10;
task{2}{1}.numBlocks = 1;
task{2}{1}.parameter.dir = 0;
task{2}{1}.parameter.coherence = 0;

% the second phase has 2 second bursts of directions, followed by 
% a top-up period of the same direction
task{2}{2}.segmin = [2 6];
task{2}{2}.segmax = [2 10];
task{2}{2}.parameter.dir = 0:60:360;
task{2}{2}.parameter.coherence = 1;
task{2}{2}.random = 1;

% initialize our task
for phaseNum = 1:length(task{2})
  [task{2}{phaseNum} myscreen] = initTask(task{2}{phaseNum},myscreen,@startSegmentCallback,@updateScreenCallback);
end

% init the stimulus
global stimulus;
myscreen = initStimulus('stimulus',myscreen);
stimulus = initDots(stimulus,myscreen);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% run the eye calibration
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
myscreen = eyeCalibDisp(myscreen);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Main display loop
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
phaseNum = 1;
while (phaseNum <= length(task{2})) && ~myscreen.userHitEsc
  % update the dots
  [task{2} myscreen phaseNum] = updateTask(task{2},myscreen,phaseNum);
  % update the fixation task
  [task{1} myscreen] = updateTask(task{1},myscreen,1);
  % flip screen
  myscreen = tickScreen(myscreen,task);
end

% if we got here, we are at the end of the experiment
myscreen = endTask(myscreen,task);
clear global stimulus;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% function that gets called at the start of each segment
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [task myscreen] = startSegmentCallback(task, myscreen)

global stimulus;
if (task.thistrial.thisseg == 1)
  stimulus.dots.coherence = task.thistrial.coherence;
else
  stimulus.dots.coherence = 0;
end
stimulus.dots.dir = task.thistrial.dir;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% function that gets called to draw the stimulus each frame
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [task myscreen] = updateScreenCallback(task, myscreen)

global stimulus
mglClearScreen;
stimulus = updateDots(stimulus,myscreen);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% function to init the dot stimulus
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function stimulus = initDots(stimulus,myscreen)

% convert the passed in parameters to real units
if ~isfield(stimulus,'dots') || ~isfield(stimulus.dots,'rmax'), stimulus.dots.rmax = 15;,end
if ~isfield(stimulus.dots,'xcenter'), stimulus.dots.xcenter = 0;,end
if ~isfield(stimulus.dots,'ycenter'), stimulus.dots.ycenter = 0;,end
if ~isfield(stimulus.dots,'dotsize'), stimulus.dots.dotsize = 3;,end
if ~isfield(stimulus.dots,'density'), stimulus.dots.density = 3;,end
if ~isfield(stimulus.dots,'coherence'), stimulus.dots.coherence = 1;,end
if ~isfield(stimulus.dots,'speed'), stimulus.dots.speed = 5;,end
if ~isfield(stimulus.dots,'dir'), stimulus.dots.dir = 0;,end

% actually a square patch of dots that get stenciled
% so calculate width and height
stimulus.dots.width = stimulus.dots.rmax*2;
stimulus.dots.height = stimulus.dots.rmax*2;

% get the number of dots
stimulus.dots.n = round(stimulus.dots.width*stimulus.dots.height*stimulus.dots.density);

% get max and min points for dots
stimulus.dots.xmin = -stimulus.dots.width/2;
stimulus.dots.xmax = stimulus.dots.width/2;
stimulus.dots.ymin = -stimulus.dots.height/2;
stimulus.dots.ymax = stimulus.dots.height/2;

% get initial position
stimulus.dots.x = rand(1,stimulus.dots.n)*stimulus.dots.width;
stimulus.dots.y = rand(1,stimulus.dots.n)*stimulus.dots.height;

% get the step size
stimulus.dots.stepsize = stimulus.dots.speed/myscreen.framesPerSecond;

% create stencil
mglClearScreen;
mglStencilCreateBegin(1);
% and draw that oval
mglGluDisk(stimulus.dots.xcenter,stimulus.dots.ycenter,[stimulus.dots.rmax stimulus.dots.rmax]/2,[1 1 1],60);
mglStencilCreateEnd;
mglClearScreen;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% function to update dot positions and draw them to screen
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function stimulus = updateDots(stimulus,myscreen)

% get the dots step
stimulus.dots.xstep = cos(pi*stimulus.dots.dir/180)*stimulus.dots.stepsize;
stimulus.dots.ystep = sin(pi*stimulus.dots.dir/180)*stimulus.dots.stepsize;

% pick a random set of dots
stimulus.dots.coherent = rand(1,stimulus.dots.n) < stimulus.dots.coherence;

% now move those dots in the right direction
stimulus.dots.x(stimulus.dots.coherent) = stimulus.dots.x(stimulus.dots.coherent)+stimulus.dots.xstep;
stimulus.dots.y(stimulus.dots.coherent) = stimulus.dots.y(stimulus.dots.coherent)+stimulus.dots.ystep;

% randomwalk rule
thisdir = rand(1,sum(~stimulus.dots.coherent))*2*pi;
stimulus.dots.x(~stimulus.dots.coherent) = stimulus.dots.x(~stimulus.dots.coherent)+cos(thisdir)*stimulus.dots.stepsize;
stimulus.dots.y(~stimulus.dots.coherent) = stimulus.dots.y(~stimulus.dots.coherent)+sin(thisdir)*stimulus.dots.stepsize;

% movshon noise
%stimulus.dots.x(~stimulus.dots.coherent) = rand(1,sum(~stimulus.dots.coherent))*stimulus.dots.width;
%stimulus.dots.y(~stimulus.dots.coherent) = rand(1,sum(~stimulus.dots.coherent))*stimulus.dots.height;

% make sure we haven't gone off the patch
% do the dots separately for left and right hand side
stimulus.dots.x(stimulus.dots.x < stimulus.dots.xmin) = stimulus.dots.x(stimulus.dots.x < stimulus.dots.xmin)+stimulus.dots.width;
stimulus.dots.x(stimulus.dots.x > stimulus.dots.xmax) = stimulus.dots.x(stimulus.dots.x > stimulus.dots.xmax)-stimulus.dots.width;
stimulus.dots.y(stimulus.dots.y < stimulus.dots.ymin) = stimulus.dots.y(stimulus.dots.y < stimulus.dots.ymin)+stimulus.dots.height;
stimulus.dots.y(stimulus.dots.y > stimulus.dots.ymax) = stimulus.dots.y(stimulus.dots.y > stimulus.dots.ymax)-stimulus.dots.height;

% draw the dots
mglStencilSelect(1);
mglPoints2(stimulus.dots.x,stimulus.dots.y,stimulus.dots.dotsize,[1 1 1]);
mglStencilSelect(0);




